File: c++/quick_sort/quick_sort.cpp
#include <iostream>
#include <stdlib.h>

using namespace std;

void fillArray(int* ary, int max, int valRange)
{
  for (int i = 0; i < max; i++) {
    ary[i] = rand() % valRange;
  }
}

void printArray(int* ary, int from, int to)
{
  int i = from;
  for (; i < to - 1; ++i) {
    cout << ary[i] << " ";
  }
  cout << ary[i] << endl;
}

int partition(int* ary, int start, int end, int pivot)
{
  int left = start - 1;
  int right = end + 1;

  while(true) {
    while(ary[++left] < pivot);
    while(right > 0 && ary[--right] > pivot);

    if (left >= right) {
      break;
    }

    swap(ary[left], ary[right]);
  }
  return left;
}

void quicksort(int* ary, int start, int end)
{
  if (end - start < 1) {
    return;
  }

  int pivot = ary[end];
  int partAt = partition(ary, start, end, pivot);
  quicksort(ary, start, partAt - 1);
  quicksort(ary, partAt + 1, end);
}


int main()
{
  srand(time(NULL));

  int aryLen = 11;
  int randRange = 100;
  int ary[aryLen] = {0};
  int pivotVal = 50;

  fillArray(ary, aryLen, randRange);
  printArray(ary, 0, aryLen);
  quicksort(ary, 0, aryLen - 1);
  printArray(ary, 0, aryLen);

  return 0;
}

File: ruby/quick_sort/quick_sort.rb
def quickSortRecurse(ary, left, right)
  if (left < right)
    pivot = ary[right]

    i = left - 1
    (left...right).each { |j|
      if ary[j] < pivot
        i += 1
        if i != j # swap i & j
          ary[i] += ary[j]
          ary[j] = ary[i] - ary[j]
          ary[i] -= ary[j]
        end
      end
    }

    pivotIndex = i + 1
    if pivotIndex != right # swap pivotIndex & right
      ary[pivotIndex] += ary[right]
      ary[right] = ary[pivotIndex] - ary[right]
      ary[pivotIndex] -= ary[right]
    end

    quickSortRecurse(ary, left, pivotIndex - 1)
    quickSortRecurse(ary, pivotIndex + 1, right)
  end
end

def quickSort(ary)
  quickSortRecurse(ary, 0, ary.size - 1)
end

aryLen = 10
ary = Array.new(aryLen) { |i| aryLen - i }
puts ary.to_s
quickSort(ary)
puts ary.to_s